Add SparkV2Filters

huaxingao opened the pull request:
This PR adds `SparkV2Filters` and `TestSparkV2Filters`. In next PR, I will make `SparkScanBuilder` implement `SupportsPushDownV2Filters` and add a test `TestV2FilteredScan`. 

Currently, Spark still uses V2 Filter in `SupportsRuntimeFiltering` and `SupportsDelete`. I have a PR to add `SupportsRuntimeV2Filtering` and will have a PR to add V2 Filter in `SupportsDelete`  in Spark. Before these support are released on Spark side, we can use `Filter.toV2` to convert V1 filter to V2 filter for now.

Spark V2 Filter format:
```
Predicate(String name, Expression[] children)
```

The Predicate could be

```
name: IS_NULL
children[0] is NamedReference

name: IS_NOT_NULL
children[0] is NamedReference

name: STARTS_WITH
children[0] is NamedReference
children[1] is LiteralValue

name: IN
children[0] is NamedReference
children[1...n] is LiteralValue

name: =
children[0] is NamedReference
children[1] is LiteralValue

name: <>
children[0] is NamedReference
children[1] is LiteralValue

name: <=>
children[0] is NamedReference
children[1] is LiteralValue

name: <
children[0] is NamedReference
children[1] is LiteralValue

name: <=
children[0] is NamedReference
children[1] is LiteralValue

name: >
children[0] is NamedReference
children[1] is LiteralValue

name: >=
children[0] is NamedReference
children[1] is LiteralValue

name: AND
  public And(Predicate left, Predicate right) {
    super("AND", new Predicate[]{left, right});
  }

name: OR
  public Or(Predicate left, Predicate right) {
    super("OR", new Predicate[]{left, right});
  }

name: NOT
  public Not(Predicate child) {
    super("NOT", new Predicate[]{child});
  }

name: ALWAYS_TRUE
  public AlwaysTrue() {
    super("ALWAYS_TRUE", new Predicate[]{});
  }

name: ALWAYS_FALSE
  public AlwaysFalse() {
    super("ALWAYS_FALSE", new Predicate[]{});
  }
```

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,356 @@
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  private static Predicate checkIfPredicateValid(Predicate predicate) {
```
I think Spark should have this check to make sure all the V2 Filters are in valid format. Otherwise, this kind of invalid [filter](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestDelete.java#L433) will get though. 

I used to have this check on Spark side, but somehow it got removed during refactor. I will add this check back. Before this is available, I will have this check here. I will remove once this is in Spark.

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,356 @@
+                convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value()));
+          } else {
+            return greaterThan(unquote(getAttrName(predicate.children()[1])),
+                convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[0]).value()));
```
I think this kind of normalization (e.g. `1 < a => a > 1`) should be done in Spark. I used to have this, but I just found out it was removed. I will add this back in Spark. But before this is available in Spark, I will have it here.

huaxingao commented:
cc @rdblue @aokolnychyi @RussellSpitzer 
Could you please take a look when you have time? Thanks!

huaxingao commented:
gentle ping @rdblue @aokolnychyi @RussellSpitzer

huaxingao commented:
also cc @flyrain @szehon-ho @kbendick

RussellSpitzer commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+
+  private SparkV2Filters() {}
+
+  private static final Map<String, Expression.Operation> FILTERS =
```
I find it kind of gross that Spark is making us deal with this, seems like this should be generated rather then us having to literally write out all the strings. Not that we have anything we can improve here. Still kind of hoping we can have some Substrait.io style solution for this kind of thing in the future.

huaxingao replied:
I think the main reason that Spark uses a String for the V2 Filter name is because Spark introduces [`GeneralScalarExpression`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/GeneralScalarExpression.java) in 3.3 and pushes down a lot of functions such as `+`, `-`, `*`, `/`, `ABS`, `LOG`, ... Spark doesn't want to have a class for each of these Expressions so it uses a more genetic way. V2 [`Predicate`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/filter/Predicate.java) is a subclass of `GeneralScalarExpression` so it also uses the field `name` for Filter name. 

I am actually wondering if any of these function push down can benefit Iceberg. It seems to me that for both the data predicate push down and partition pruning, Iceberg only supports predicate `attri cmp lit`, so we can't really take advantage of any of these new function push down in Spark. In Spark built-in file source such as Parquet, partition pruning uses Catalyst Expression so Spark can prune partition such as `partCol1 + partCol2 > 3`.

kbendick replied:
Agreed. But I see why it's done.

The V2FunctionCatalog works the same way. Notice the `bucket` v2 function transform opt works the same way. With a string typed `numCols` and a string typed `bucket` function that Spark searches for on the implementing library's side. The name technically doesn't matter for the field but it 100% does for the function.

https://github.com/apache/spark/blob/126870eddc57fff7408942897dd59a2c19c91c4a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/V2ExpressionUtils.scala#L90-L100

I hope @huaxingao is right though that we will be able to use Spark to parse more expressions this way though =)

RussellSpitzer commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+  @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:MethodLength"})
+  public static Expression convert(Predicate predicate) {
+    if (checkIfPredicateValid(predicate) == null) {
+      return null;
```
I'm a little confused why we use the return value here to signal error. The Code above iterates through the predicates one at a time so we don't collect all of the "non-convertable" predicates, I could see if we wanted to generate a large error message with all the filters we couldn't convert but it seems like we could just assert here rather than in the parent "convert" method?

huaxingao replied:
The idea is that if the Spark V2 filter is not valid, then we simply return a `null` and skip the convert code, then later on we don't need to bind and push down this predicate (https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L120)

RussellSpitzer replied:
But we throw an exception here? Shouldn't the behavior just be to ignore the predicate then?
https://github.com/apache/iceberg/pull/5302/files#diff-d21a6d1f3fb11bbc8d5e9ced54b7fd025e04a70a54c8263ae49313ea3cb274beR84-R85

huaxingao replied:
I see what you mean now. Sorry I wasn't realize that you were talking about the convert method that takes the `Predicate` array.

This method `convert(Predicate predicate)` will be used by [`SparkScanBuilder`.`pushPredicates`](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L111) directly. If the Spark v2 filter can't be converted to Iceberg Expression, we just ignore the filter and not push it down.

The method `convert(Predicate[] predicates)` will be used by [`SparkWriteBuilder`.`overwrite`](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkWriteBuilder.java#L122) and [`SparkTable`.`deleteWhere`](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java#L311). I follow the same logic as [`SparkFilter.convert(Filter[] filters)`](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java#L311). By definition, `overwrite` can only replace data iff all the filters match and `deleteWhere` can only delete rows iff all the filter expressions match. That's why we throw Exception in `convert(Predicate[] predicates)` if any of the filter can't be converted to Iceberg Expression.

huaxingao replied:
I think this over, I probably should call `convert(Predicate predicate)` one by one in [SparkWriteBuilder.overwrite](https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkWriteBuilder.java#L122). If I get null, then throw Exception saying `Cannot overwrite files using filters because not all the filter[s] can be converted correctly`. This is probably better than the error message `Cannot convert predicate to Iceberg`.

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  private static Predicate checkIfPredicateValid(Predicate predicate) {
```
I think we usually prefer shorter method names, so maybe valid(Predicate predicate) is better?  Also, kind of related to the previous comment, why do we have the return value be a boolean instead of predicate?  (valid seems to be a true/false question)

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+        case GT:
+        case GT_EQ:
+        case EQ:
+          if (predicate.children().length != 2
```
Just a personal preference but maybe its easier to read (if we go with the boolean return value approach):
```
if (predicate.children.length() != 2) {
  return false;
}
return (predicate.children[0] instanceof NamedReference && predicate.children[1] instanceof LiteralReference) ||
(predicate.children[0] instanceof LiteralRefernce && predicate.children[1] instanceof NamedReference]);
```

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+          if (!(predicate.children()[0] instanceof NamedReference)) {
+            return null;
+          } else {
+            for (int i = 1; i < predicate.children().length; i++) {
```
Also optional style suggestion, can use predicate.children.stream().allMatch()

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+
+        case AND:
+          And andFilter = (And) predicate;
+          if (checkIfPredicateValid(andFilter.left()) == null
```
Same if we change, we could have an easier to read:
return valid(andFilter.left) && valid(andFilter.right);

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+          if (predicate.children()[1] instanceof LiteralValue) {
+            return lessThan(
+                unquote(predicate.children()[0].toString()),
+                convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value()));
```
I'm not very familiar but In the old SparkFilters code there used to be a 'convertLiteral' functionality.  Do we need that for literals?

huaxingao replied:
We don't need this `convertLiteral` any more. This is the conversion that V2 filter is trying to avoid.

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+          }
+
+        case EQ: // used for both eq and null-safe-eq
+          Object value = null;
```
Nit: looks like we dont need null and "" (java defaults)

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+
+        case IN:
+          Object[] inValues = new Object[predicate.children().length - 1];
+          for (int i = 1; i < predicate.children().length; i++) {
```
Why not just make this a stream (as the next line anyway converts it to stream).  Ie, predicate.children.stream().map(c -> convertUTF8StringIfNecessary(...))

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,375 @@
+            // col NOT IN (1, 2) in Spark is equivalent to notNull(col) && notIn(col, 1, 2) in
+            // Iceberg
+            Object[] notInValues = new Object[childFilter.children().length - 1];
+            for (int i = 1; i < childFilter.children().length; i++) {
```
Same comment (use a stream)

szehon-ho reviewed (commented):
Hi @huaxingao I guess most of the code is direct from SparkFilters.java, I did an early code pass.

It would be really nice if Spark did the normalization like in the previous version (the order of comparison operations), so the code mostly stays the same as before.

huaxingao commented:
@szehon-ho Thank you very much for taking time to review. I have addressed the comments. Could you please take another look when you are free? Thanks a lot!

szehon-ho commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,353 @@
+              return handleEqual(unquote(attributeName), value);
+            }
+          }
+          break;
```
Nit: maybe we should return null explicitly like the other case instead of break.  Also curious, will this ever happen?

huaxingao replied:
This will not happen. I removed this.

szehon-ho commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkV2Filters.java:
```diff
@@ -0,0 +1,215 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestSparkV2Filters {
```
Do we need some test of the cases of the expression the other way ( literal, attribute)

huaxingao replied:
Added some tests. Thanks!

kbendick replied:
Nit / non-blocking / for follow up if desired: We might break these up into two or three sets of tests.

I can see:
1. Literal to Attribute
2. Attribute to Literal
3. Invalid assertions.

This could totally be done in a follow up PR, but because the stringly typed nature of these areas is somewhat error prone in my opinion I think we should try to make the tests extensible. But not important for this PR imo (which has been around long enough and would be good to be using).

huaxingao replied:
I will refactor the tests in a follow up. Thanks for the suggestion!

szehon-ho reviewed (commented):
Looks better to me , thanks @huaxingao .  Just curious, any chance we can avoid having the normalization code in Iceberg, or will it take too long as we need to wait for a Spark pr and release?

huaxingao commented:
> any chance we can avoid having the normalization code in Iceberg, or will it take too long as we need to wait for a Spark pr and release?

I have already added the normalization code in Spark, but we need to wait for 3.4 release.

szehon-ho reviewed (approved):
Cool, looks good to me.  May wait a little to see if anyone else wants to take a look, maybe @aokolnychyi

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,352 @@
+          }
+
+        case LT_EQ:
+          if (predicate.children()[1] instanceof LiteralValue) {
```
Nit / non-blocking style question: Would it make sense to use named constant for the ordinals? They're used quite a lot.

Maybe the following?
```
private static int FIRST_PREDICATE_CHILD_ORDINAL = 0;
private static int SECOND_PREDICATE_CHILD_ORGINAL = 1;
```

Not tied to the naming but Anton suggested I do similar in the `truncate` v2 function PR and it turned out nicely. Might be overkill since they don't correspond to much besides first and second though. 

https://github.com/apache/iceberg/blob/b3062e443848c9badd7332c4b26c14f2b257cc4a/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java#L56-L57

kbendick replied:
Again, this might be overkill as they don't correspond much to anything meaningful (like the ones above in `TruncateFunction` do).

huaxingao replied:
Fixed, but made the names a little shorter.

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,352 @@
+    return value;
+  }
+
+  private static Expression handleEqual(String attribute, Object value) {
```
Maybe just `NaNSafeEquals` or `SafeEquals`. This is fine as well.

kbendick replied:
Edit - Nevermind this is lifted from the original SparkFilters. Leaving it as is is the best choicee imo.

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,352 @@
+          if (predicate.children().length != 2) {
+            return false;
+          }
+          return (predicate.children()[0] instanceof NamedReference
+                  && predicate.children()[1] instanceof LiteralValue)
```
This pattern comes up enough (and is somewhat confusing to read).

Maybe `binaryNamedReferenceComparedToLiteral` and `binaryLiterComparedToReference`? Though I guess if Spark is moving towards inlining with strings etc, it might be better to not try to deviate from that at first.

We can always refactor so I'd save something like this for a follow up PR (if at all). We'd probably wind up building up _a lot_ of functions.

huaxingao replied:
Fixed, but also made the names a little shorter.

kbendick commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkV2Filters.java:
```diff
@@ -0,0 +1,254 @@
+  @Test
+  public void testNestedInInsideNot() {
+    NamedReference namedReference1 = FieldReference.apply("col1");
+    LiteralValue v1 = new LiteralValue(1, IntegerType$.MODULE$);
```
Nit: I think `IntegerType$.Module$` is equivalent to `DataTypes.IntegerType`, is it not? They might be one dollar sign apart 🙃  but if they're equivalent we should use that.

huaxingao replied:
Fixed. Thanks

kbendick reviewed (approved):
Overall this looks good to me.

There are some nits that I would maybe change style wise, but I don't think it's worth holding up the PR necessarily. It's also just my opinion and not required.

Also, this should eventually cover at least the same tests as the V1 filters, but if that's already been met than that's fine.

Thanks @huaxingao! And thanks for your patience on the reviews!

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,352 @@
+
+  private static final Map<String, Expression.Operation> FILTERS =
+      ImmutableMap.<String, Expression.Operation>builder()
+          .put("ALWAYS_TRUE", Expression.Operation.TRUE)
+          .put("ALWAYS_FALSE", Expression.Operation.FALSE)
+          .put("=", Expression.Operation.EQ)
+          .put("<=>", Expression.Operation.EQ)
+          .put(">", Expression.Operation.GT)
+          .put(">=", Expression.Operation.GT_EQ)
+          .put("<", Expression.Operation.LT)
+          .put("<=", Expression.Operation.LT_EQ)
+          .put("IN", Expression.Operation.IN)
+          .put("IS_NULL", Expression.Operation.IS_NULL)
+          .put("IS_NOT_NULL", Expression.Operation.NOT_NULL)
+          .put("AND", Expression.Operation.AND)
+          .put("OR", Expression.Operation.OR)
+          .put("NOT", Expression.Operation.NOT)
+          .put("STARTS_WITH", Expression.Operation.STARTS_WITH)
```
I have added named constants for the predicates.

huaxingao commented:
Thanks a lot @kbendick for taking time reviewing the PR! I have addressed the comments. Could you please take one more look when you are free? Thanks!

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+
+  private static final Pattern BACKTICKS_PATTERN = Pattern.compile("([`])(.|$)");
+
+  private SparkV2Filters() {}
```
Nit: Maybe move the constructor before or after all of the private static constants?

huaxingao replied:
Fixed. Thanks!

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+          if (predicate.name().equals("=")) {
+            // comparison with null in normal equality is always null. this is probably a mistake.
+            Preconditions.checkNotNull(
+                value, "Expression is always false (eq is not null-safe): %s", predicate);
```
Question: will this fail the query?

So if somebody writes `select a, b from foo where c = null`, the query will throw? Would we be better off logging a warning and then using `Expressions.alwaysFalse()` (or null), or is throwing more user friendly since probably they didn’t mean for that to be the query — the query basically will not match anything right?

Genuinely asking as I’m not sure if people use this pattern at all. You’re right it seems like a mistake.

Maybe we should add a unit test that this throws the specified exception?

I’m genuinely not sure which is best so just asking. I’m good with it as is.

huaxingao replied:
This is actually borrowed from `SparkFilters`

kbendick replied:
Ah then absolutely keep it :)

aokolnychyi replied:
It is unexpected and I think we wrap the conversion logic into try-catch, so we should be fine.

kbendick commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+          .build();
+
+  private static final int FIRST_ORDINAL = 0;
+  private static final int SECOND_ORDINAL = 1;
```
I know I suggested first_ordinal and second_ordinal, but if you’re not in love with those, after looking at it further I think “left child” and “right child” might make sense too.

But I also don’t see any need to bike shed / change further just for the sake of changing. This is plenty readable as-is for me. Just throwing it out there as another idea.

huaxingao replied:
I took @aokolnychyi's suggestion to use `leftChild` and `rightChild`

kbendick reviewed (approved):
This looks good to me. Left a few comments, but I don’t think they need to be addressed — just last minute other options I thought of in case you decide you don’t want to use something as it currently is.

But this looks good to me. I can see why Spark doesn’t want so many classes but it does add a little more burden on us. I think you did a really good job with it considering. Thanks @huaxingao!

aokolnychyi commented:
I'd like to take a look tomorrow or on Friday as well.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+
+  @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:MethodLength"})
+  public static Expression convert(Predicate predicate) {
+    if (!valid(predicate)) {
```
@huaxingao, can you elaborate a bit on what is considered invalid in this case? Are we trying to validate the Spark logic or are we trying to check if the filter is supported by Iceberg?

huaxingao replied:
Mainly trying to validate the Spark filter logic. In binary comparison predicate, Expression cmp Expression is also valid in Spark, but I think Iceberg only accepts Attribute cmp Literal, so I also make sure it's either attribute cmp lit or lit cmp attribute.

aokolnychyi replied:
I see. It makes sense but what do you think about adapting the conversion logic a bit instead of doing two passes? We anyway need to check if left or right child is literal/reference. You could define helper methods like this.

```
private static boolean isRef(org.apache.spark.sql.connector.expressions.Expression expr) {
  return expr instanceof NamedReference;
}

private static boolean isLiteral(org.apache.spark.sql.connector.expressions.Expression expr) {
  return expr instanceof Literal;
}
```

huaxingao replied:
Changed. Thanks for the suggestion!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+          return null;
+
+        case AND:
+          {
```
nit: Do we need these extra `{}` here and 2 branches below?

huaxingao replied:
This is required. Otherwise, I can't use the same variable names (`left` and `right`) in difference case blocks.

aokolnychyi reviewed (commented):
This is a great effort. Thanks for doing this, @huaxingao!

I think we need to remove some of the workarounds we had for V1 filters like backticks unquoting and just use field name arrays if possible.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+          return Expressions.alwaysFalse();
+
+        case IS_NULL:
+          return isNull(unquote(predicate.children()[FIRST_ORDINAL].toString()));
```
Weird. I think my original comment wasn't posted.

I was saying that we shouldn't use unquoting workarounds we had for V1 filters. Instead, we should use field arrays in V2 predicates.

I'd consider something like this (needs to be polished further).

```
private static final Joiner DOT = Joiner.on(".");

...

@SuppressWarnings("unchecked")
private static <T> T child(Predicate predicate, int index) {
  return (T) predicate.children()[index];
}

@SuppressWarnings("unchecked")
private static <T> T child(Predicate predicate) {
  org.apache.spark.sql.connector.expressions.Expression[] children = predicate.children();
  Preconditions.checkArgument(children.length == 1, "TBD");
  return (T) children[0];
}

private static String toColumnName(NamedReference ref) {
  return DOT.join(ref.fieldNames());
}
```

And using it like this:

```
...

case IS_NULL:
  return isNull(toColumnName(child(predicate)));

case NOT_NULL:
  return notNull(toColumnName(child(predicate)));

...
```

kbendick replied:
Oh big +1 I like this a lot.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+          return notNull(unquote(predicate.children()[FIRST_ORDINAL].toString()));
+
+        case LT:
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
```
Isn't `LiteralValue` private to Spark? Should we use the public interface?

huaxingao replied:
Changed to `Literal`. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+    return null;
+  }
+
+  private static Object convertUTF8StringIfNecessary(Object value) {
```
Can this method accept `Literal` from Spark and produce the Java object?

huaxingao replied:
Fixed. Thanks!

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+  private static final String STARTS_WITH = "STARTS_WITH";
+
+  private static final Map<String, Expression.Operation> FILTERS =
+      ImmutableMap.<String, Expression.Operation>builder()
```
Fixed. Thanks!

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,385 @@
+  private static final int FIRST_ORDINAL = 0;
+  private static final int SECOND_ORDINAL = 1;
+
+  public static Expression convert(Predicate[] predicates) {
```
Removed.

kbendick commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkV2Filters.java:
```diff
@@ -0,0 +1,252 @@
+  public void testNotIn() {
+    NamedReference namedReference = FieldReference.apply("col");
+    LiteralValue v1 = new LiteralValue(1, DataTypes.IntegerType);
+    LiteralValue v2 = new LiteralValue(2, DataTypes.IntegerType);
```
Nit: Can `Literal` be used in place of `LiteralValue` here?

huaxingao replied:
Good catch! Will fix it.

huaxingao replied:
Oh actually `Literal` is an abstract class, will have to use `LiteralValue` here.

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,380 @@
+  @SuppressWarnings("unchecked")
+  private static <T> T child(Predicate predicate) {
+    org.apache.spark.sql.connector.expressions.Expression[] children = predicate.children();
+    Preconditions.checkArgument(children.length == 1, "%s should have one child", predicate);
```
Will fix. Thanks

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+          return Expressions.alwaysFalse();
+
+        case IS_NULL:
+          return isNull(toColumnName(child(predicate)));
```
To have this as one pass, we could have:

```
return isRef(child(predicate)) ? isNull(toColumnName(child(predicate))) : null;
```

huaxingao replied:
Fixed. Thanks for the suggestion!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+          return notNull(toColumnName(child(predicate)));
+
+        case LT:
+          if (rightChild(predicate) instanceof Literal) {
```
To have this as one pass, we could have:

```
if (isRef(leftChild(predicate)) && isLiteral(rightChild(predicate))) {
  String columnName = toColumnName(leftChild(predicate));
  return lessThan(columnName, convertLiteral(rightChild(predicate)));

} else if (isRef(rightChild(predicate)) && isLiteral(leftChild(predicate))) {
  String columnName = toColumnName(rightChild(predicate));
  return greaterThan(columnName, convertLiteral(leftChild(predicate)));

} else {
  return null;
}
```

huaxingao replied:
Fixed. Thanks for the suggestion!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+    return null;
+  }
+
+  private static Object convertUTF8StringIfNecessary(Literal literal) {
```
nit: Just `convertLiteral`?

huaxingao replied:
Changed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+          }
+
+        case IN:
+          return in(
```
To have this as one pass, we could have:

```
if (isSupportedInPredicate(predicate)) {
  return in(
      toColumnName((NamedReference) predicate.children()[0]),
      Arrays.stream(predicate.children())
          .skip(1)
          .map(child -> convertLiteral(((Literal<?>) child)))
          .filter(Objects::nonNull)
          .collect(Collectors.toList()));
} else {
  return null;
}
```

huaxingao replied:
Changed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+          if (childFilter.name().equals("IN")) {
+            // infer an extra notNull predicate for Spark NOT IN filters
+            // as Iceberg expressions don't follow the 3-value SQL boolean logic
+            // col NOT IN (1, 2) in Spark is equivalent to notNull(col) && notIn(col, 1, 2) in
```
nit: What about replacing `equivalent` with `equal` so that the last line of the comment would fit on one line?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+                  .collect(Collectors.toList()));
+
+        case NOT:
+          Not notFilter = (Not) predicate;
```
nit: `notPredicate`?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+
+        case NOT:
+          Not notFilter = (Not) predicate;
+          Predicate childFilter = notFilter.child();
```
nit: `childPredicate`?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+        case NOT:
+          Not notFilter = (Not) predicate;
+          Predicate childFilter = notFilter.child();
+          if (childFilter.name().equals("IN")) {
```
nit: Use `IN` constant defined above?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+                    toColumnName(leftChild(childFilter)),
+                    Arrays.stream(childFilter.children())
+                        .skip(1)
+                        .map(val -> convertUTF8StringIfNecessary(((Literal) val)))
```
nit: `Literal<?>` to avoid warnings about raw usage of parameterized types?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+        case STARTS_WITH:
+          return startsWith(
+              toColumnName(leftChild(predicate)),
+              convertUTF8StringIfNecessary(rightChild(predicate)).toString());
```
Hm, do we need `toString` or a cast would work too to be sure we have a string?

huaxingao replied:
Seems this `toString` is needed.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+    return false;
+  }
+
+  private static boolean namedRefCompLit(Predicate predicate) {
```
I think splitting this into `isRef` and `isLiteral` should allow us to combine the validation and conversion into one pass. Then we won't have to jump between methods when looking at the code.

huaxingao replied:
Agree!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+            value = convertUTF8StringIfNecessary(leftChild(predicate));
+          }
+
+          if (predicate.name().equals("=")) {
```
nit: Shall we use `EQ` constant defined above?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+    return null;
+  }
+
+  private static Object convertUTF8StringIfNecessary(Literal literal) {
```
nit: `Literal` -> `Literal<?>`

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+            return false;
+          }
+          return namedRefCompLit(predicate)
+              && ((Literal<?>) rightChild(predicate)).value() instanceof String;
```
Hm, was it actually correct? Wouldn't it be `UTF8String`?

huaxingao replied:
Good catch! It should be `UTF8String`

aokolnychyi commented on spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/TestSparkV2Filters.java:
```diff
@@ -0,0 +1,252 @@
+public class TestSparkV2Filters {
+
+  @Test
+  public void testV2Filters() {
```
Can we add a test for `startsWith`? I think the was a problem in the conversion and tests did not catch it.

huaxingao replied:
Added. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  private static boolean valid(Predicate predicate) {
```
I kept jumping from `convert` to `valid` when looking at the code. I feel having both the validation and conversion in one function would be a bit easier.

huaxingao replied:
Agree!

aokolnychyi reviewed (commented):
I did a detailed pass. I think this PR is really close. I'd probably combine the validation and conversion into one method. It also seems we have a bug for `starts_with`.

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+
+  @SuppressWarnings("unchecked")
+  private static <T> T leftChild(Predicate predicate) {
+    org.apache.spark.sql.connector.expressions.Expression[] children = predicate.children();
```
Added `childAtIndex`. Thanks for the suggestion!

huaxingao commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,381 @@
+  private SparkV2Filters() {}
+
+  @SuppressWarnings("unchecked")
+  private static <T> T child(Predicate predicate) {
```
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,345 @@
+
+  @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:MethodLength"})
+  public static Expression convert(Predicate predicate) {
+
```
nit: This empty line is probably unnecessary.

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,345 @@
+        case NOT:
+          Not notPredicate = (Not) predicate;
+          Predicate childPredicate = notPredicate.child();
+          if (childPredicate.name().equals(IN)) {
```
Shall we add also `isSupportedInPredicate(childPredicate)` to the condition?

huaxingao replied:
Added. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,345 @@
+    if (!isRef(childAtIndex(predicate, 0))) {
+      return false;
+    } else {
+      return Arrays.stream(predicate.children()).skip(1).allMatch(val -> isLiteral(val));
```
nit: `val -> isLiteral(val)` -> `SparkV2Filters::isLiteral`?

huaxingao replied:
Fixed. Thanks!

aokolnychyi commented on spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java:
```diff
@@ -0,0 +1,345 @@
+          }
+
+        case STARTS_WITH:
+          if (isRef(leftChild(predicate))
```
Is it realistic that Spark can pass something other than a reference followed by String? I think the other validation in this method is reasonable as we could technically have other valid expressions with structs inside IN or other predicates, which Iceberg does not support yet. I am not sure about this one, though.

I am fine keeping it too but I'd move it to `isSupportedStartsWithPredicate`. Up to you here, @huaxingao.

aokolnychyi reviewed (approved):
Looks great! A few final nits and is ready to go! Thanks for the work, @huaxingao!

aokolnychyi commented:
Thanks, @huaxingao! Thanks for reviewing, @szehon-ho @kbendick!

huaxingao commented:
Thank you all very much for your help!! @aokolnychyi @szehon-ho @kbendick
